New entry Multilevel Inheritance (#6049)
Codecademy/docs: content/cpp/concepts/inheritance/terms/multilevel-inheritance/multilevel-inheritance.md

---
Title: 'Multilevel Inheritance'
Description: 'Implements a chain of inheritance where a derived class inherits from another derived class, forming a parent-child-grandchild relationship between classes.'
Subjects:
  - 'Computer Science'
  - 'Code Foundations'
Tags:
  - 'Inheritance'
  - 'OOP'
  - 'Classes'
  - 'Objects'
CatalogContent:
  - 'learn-c-plus-plus'
  - 'paths/computer-science'
---

> **Multilevel inheritance** is an Object-Oriented Programming (OOP) concept where a class can inherit properties and methods from a class that is already inherited from another class, creating a hierarchy of classes. This forms a parent-child-grandchild relationship between classes, allowing for the creation of specialized classes based on existing ones.

Reviewer: Add docs link to OOP concepts - 

```suggestion
**Multilevel inheritance** is an [Object-Oriented Programming (OOP)](https://www.codecademy.com/resources/docs/general/programming-paradigms/object-oriented-programming) concept where a class can inherit properties and methods from a class that is already inherited from another class, forming a hierarchical class structure. This forms a parent-child-grandchild relationship between classes, enabling the creation of specialized classes from existing ones.
```

## Syntax

> In C++, multilevel inheritance follows this general syntax:

Reviewer: ```suggestion
In C++, multilevel inheritance follows this syntax:
```

> ```cpp

Reviewer: We use `pseudo` code blocks for syntax -

```suggestion
```pseudo
```

class BaseClass {
    // Base class members
};

class DerivedClass1 : public BaseClass {
    // First level derived class members
};

class DerivedClass2 : public DerivedClass1 {
    // Second level derived class members
};
```

## Example

> Simple example to demonstrate multilevel inheritance in C++

Reviewer: This can be better written as following - 

```suggestion
Here's an example demonstrating multilevel inheritance in C++:
```

        std::cout << "Name: " << name << "\nID: " << id << std::endl;
    }
};

class Developer : public Employee {
protected:
    std::string programmingLanguage;

public:
    Developer(std::string name, int id, std::string lang)
        : Employee(name, id), programmingLanguage(lang) {}

    void displayInfo() override {
        Employee::displayInfo();
        std::cout << "Role: Developer\n";
        std::cout << "Programming Language: " << programmingLanguage << std::endl;
    }
};

class SeniorDeveloper : public Developer {
private:
    int teamSize;
    std::string projectName;

public:
    SeniorDeveloper(std::string name, int id, std::string lang,
                    int team, std::string project)
        : Developer(name, id, lang), teamSize(team), projectName(project) {}

    void displayInfo() override {
        Developer::displayInfo();
        std::cout << "Position: Senior Developer\n";
        std::cout << "Team Size: " << teamSize << "\n";
        std::cout << "Current Project: " << projectName << std::endl;
    }
};

int main() {
    SeniorDeveloper lead("Alice Johnson", 1001, "C++", 5, "Payment Gateway");
    lead.displayInfo();
    return 0;
}
```

> This example demonstrates how multilevel inheritance allows a class to inherit features from multiple levels of parent classes, creating a clear and organized hierarchy of related classes. The C++ implementation includes additional features like virtual functions for proper polymorphic behavior and access specifiers (public, protected, private) to control member accessibility.

Reviewer: Add relevant doc links -

```suggestion
This example demonstrates how multilevel inheritance allows a class to inherit features across multiple levels, forming a structured class hierarchy. The implementation includes virtual functions for proper [polymorphic behavior](https://www.codecademy.com/resources/docs/cpp/polymorphism) and access specifiers (`public`, `protected`, `private`) to control member accessibility.
```

Reviewer: We should move this above the code block - we try to keep the explanation very short for codebyte examples and they are the last part of the entry.
